src/write.rs

```diff
@@ -713,4 +971,170 @@ mod tests {
+            Some(Op::Sub(1)),
+            Some(Op::Add(1)),
+        ]);
+        let mut rev_dirty_range = 0..0;
+        for op in [Op::Add(1), Op::Sub(1), Op::Add(1)] {
+            w.compress_insert_op(op, &mut rev_dirty_range);
+        }
+        // rev_dirty_range.end is 3 instead of 2 because when inserting last Add at rev_idx: 3 it can't know about Sub at rev_idx: 2.
+        assert_eq!(rev_dirty_range, 2..3);
+        w.oplog
+            .iter()
+            .zip([Some(Op::Add(8)), Some(Op::Sub(2)), None, None])
+            .for_each(|(op, expected)| assert_eq!(*op, expected));
+        w.oplog.retain(Option::is_some);
+        w.oplog
+            .iter()
+            .zip([Some(Op::Add(8)), Some(Op::Sub(2))])
+            .for_each(|(op, expected)| assert_eq!(*op, expected));
+    }
+    #[test]
+    fn oplog_remove_nones_early_stop() {
+        type Op = CompressibleCounterOp<{ usize::MAX }>;
+        let (mut w, _r) = crate::new::<i32, Op>();
+        // Get first optimization out of the picture
+        w.publish();
+        assert_eq!(w.first, false);
+        // Force contrived oplog which causes none removal to stop early after failing to find a Some to swap a None with.
+        w.oplog
+            .extend([Some(Op::Add(3)), Some(Op::Sub(1)), None, Some(Op::Sub(1))]);
+        w.append(Op::Add(1));
+        w.oplog
+            .iter()
+            .zip([Some(Op::Add(4)), Some(Op::Sub(1)), Some(Op::Sub(1))])
+            .for_each(|(op, expected)| assert_eq!(*op, expected));
+    }
+    #[quickcheck]
+    fn compress_correct(mut input: Vec<i32>) -> bool {
+        // To avoid over/underflow during arithmetic
+        for x in input.iter_mut() {
+            *x &= 0b11;
+        }
+        type Op = CompressibleCounterOp<2>;
+        let (mut w, r) = crate::new::<i32, Op>();
+        // Get non-compressing first optimization out of the picture
+        w.publish();
+        assert_eq!(w.first, false);
+        // Map numbers to Ops, insert and publish them
+        let mut iter = input.chunks(8);
+        while let Some(chunk) = iter.next() {
+            // First extend while not compressing for more corner cases
+            w.oplog.extend(chunk.iter().map(|&x| {
+                Some(if x > 0 {
+                    Op::Add(x)
+                } else if x < 0 {
+                    Op::Sub(-x)
+                } else {
+                    Op::Set(0)
+                })
+            }));
+            if let Some(chunk) = iter.next() {
+                w.extend(chunk.iter().map(|&x| {
+                    if x > 0 {
+                        Op::Add(x)
+                    } else if x < 0 {
+                        Op::Sub(-x)
+                    } else {
+                        Op::Set(0)
+                    }
+                }));
+            }
+            w.publish();
+        }
+        // to avoid some macro complications
+        let val = *r.enter().unwrap();
+        // Check if value correct
```

Comment:
As you make the input logic more complex, I recommend computing the expected final value as you go, rather than trying to stitch it together from the input at the end. I suspect it'll ultimately lead to simpler code. But that's a matter of taste — either way is fine.

Reply:
Done: b852b9ceb2600c78c8789f2a91ff46c4a4f876ec